superset/explore/commands/get.py

```diff
@@ -0,0 +1,172 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import logging
+from abc import ABC
+from typing import Any, cast, Dict, Optional
+
+import simplejson as json
+from flask import current_app as app
+from flask_babel import gettext as __, lazy_gettext as _
+from sqlalchemy.exc import SQLAlchemyError
+
+from superset import db, security_manager
+from superset.commands.base import BaseCommand
+from superset.connectors.base.models import BaseDatasource
+from superset.connectors.connector_registry import ConnectorRegistry
+from superset.connectors.sqla.models import SqlaTable
+from superset.datasets.commands.exceptions import DatasetNotFoundError
+from superset.exceptions import SupersetException
+from superset.explore.commands.parameters import CommandParameters
+from superset.explore.exceptions import DatasetAccessDeniedError, WrongEndpointError
+from superset.explore.form_data.commands.get import GetFormDataCommand
+from superset.explore.form_data.commands.parameters import (
+    CommandParameters as FormDataCommandParameters,
+)
+from superset.explore.permalink.commands.get import GetExplorePermalinkCommand
+from superset.explore.permalink.exceptions import ExplorePermalinkGetFailedError
+from superset.models.sql_lab import Query
+from superset.utils import core as utils
+from superset.views.utils import (
+    get_datasource_info,
+    get_form_data,
+    sanitize_datasource_data,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class GetExploreCommand(BaseCommand, ABC):
+    def __init__(
+        self,
+        params: CommandParameters,
+    ) -> None:
+        self._actor = params.actor
+        self._permalink_key = params.permalink_key
+        self._form_data_key = params.form_data_key
+        self._dataset_id = params.dataset_id
+        self._dataset_type = params.dataset_type
+        self._slice_id = params.slice_id
+
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    def run(self) -> Optional[Dict[str, Any]]:
+        initial_form_data = {}
+
+        if self._permalink_key is not None:
+            command = GetExplorePermalinkCommand(self._actor, self._permalink_key)
+            permalink_value = command.run()
+            if permalink_value:
+                state = permalink_value["state"]
+                initial_form_data = state["formData"]
+                url_params = state.get("urlParams")
+                if url_params:
+                    initial_form_data["url_params"] = dict(url_params)
+            else:
+                raise ExplorePermalinkGetFailedError()
+        elif self._form_data_key:
+            parameters = FormDataCommandParameters(
+                actor=self._actor, key=self._form_data_key
+            )
+            value = GetFormDataCommand(parameters).run()
+            initial_form_data = json.loads(value) if value else {}
+
+        message = None
```

Comment:
Many exceptions can be caught in the `self.validate` method.

Reply:
Avoiding refactors at this stage 😉

Reply:
I agree with @zhaoyongjie

Reply:
I also agree with @zhaoyongjie. I was just trying to avoid refactors at this stage and preserve the old endpoint code but it seems it will not be possible :man_shrugging:. I'll implement the suggestion.

Reply:
Ok. I'll leave the refactor for a follow-up then. @zhaoyongjie commented that we may have parts that are not even being used anymore. After we complete the SPA work, we'll have more knowledge of the domain as well, which will give us more information for the refactoring.